### DISCUSSION TITLE
micronaut validation doesn't seem to working, is it my setup or a bug?

### DISCUSSION THREAD

**Discussion by Jonathan-Zollinger** [micronaut-projects/micronaut-core#10449] [Q&A]
## Summary 
Regex and not-blank annotations fail to... fail?

I've a POJO called 'place' which has two constructors, each with a unique regex pattern validation and `@not-blank` annotation. I've written tests to validate bad input triggers validation exceptions, but it seems any string I pass is accepted in the constructor.

I imagine I'm doing something wrong, but I dont know where I've gone wrong. Can anyone shed some light on this for me?

<details><summary>background</summary><ul><br>

ultimately this is to support a new endpoint I'm querying in my [Threshr project](https://github.com/Graqr/Threshr). the regex mocks the validation from the third party api.

I've run into this same sort of thing when I added a validator for my `tcin` class in this same repo, but instead of pushing to figure out what's wrong I added a home-grown validation in the constructor.

```java
    public void setTcins(String... tcins) throws ThreshrException {
        Predicate<String> tcinCheck = t -> !String.valueOf(t).matches("\\d{8}");
        String badTcin = Arrays.stream(tcins)
                .filter(tcinCheck)
                .collect(Collectors.joining(", "));
        if (!badTcin.isEmpty()) {
            throw new ThreshrException( "the following are invalid tcin:" + badTcin);
        }
        this.tcins = tcins;
```
</ul></details>

## Setup

I found this when creating an httpclient but have created a [repo dedicated to showing this issue](https://github.com/Jonathan-Zollinger/validation_bug). 

<details><summary>quick setup</summary><ul>

#### ● gh cli
```shell
gh repo clone Jonathan-Zollinger/validation_bug
```
#### ● plain ol' git
```shell
git clone https://github.com/Jonathan-Zollinger/validation_bug.git
```

</ul></details>

in `src/main/java/com/jonathan_zollinger/Place.java`: 
```java
@Serdeable
@Data
public class Place {
    private final String place;

    public Place(@Pattern(regexp = "^\\d{5}(-|\\s*)?(\\d{4})?$") //allows standard 5-digit zip as well as "" + 4 versions
                 @NotBlank String zipcode){
        place = zipcode;
    }
    public Place(
            @Pattern(regexp = "^([a-zA-Z\\u0080-\\u024F]+(?:. |-| |'))*[a-zA-Z\\u0080-\\u024F]*$")
            @NotBlank String city,
            @Pattern(regexp = "^[A-Z][a-z]+(?: +[A-Z][a-z]+)*$")
            @NotBlank String state){
        place = String.format("%s, %s", city, state);
    }
}
```

[its spock tests](https://github.com/Jonathan-Zollinger/validation_bug/blob/main/src/test/groovy/com/jonathan_zollinger/PlaceSpec.groovy) supply invalid args one way or another but dont trigger any errors. I've ran the tests in [gh-actions to share the error messages](https://github.com/Jonathan-Zollinger/validation_bug/actions/runs/7768809982/job/21187169460#step:4:1967), but they're all very similar to the one below

```sh
Error:  com.jonathan_zollinger.PlaceSpec.test creating a Place object with " " and "Maryland" throws an error -- Time elapsed: 0.001 s <<< FAILURE!
Expected exception of type 'jakarta.validation.ValidationException', but no exception was thrown
	at org.spockframework.lang.SpecInternals.checkExceptionThrown(SpecInternals.java:84)
	at org.spockframework.lang.SpecInternals.thrownImpl(SpecInternals.java:71)
	at com.jonathan_zollinger.PlaceSpec.test creating a Place object with "#badCity" and "#badState" throws an error(PlaceSpec.groovy:61)
```

**Comment by graemerocher** [comment]
Micronaut doesn't modify the byte code of you classes, if you want validation you have to activate it a method argument in the logic place where the type is validated, for example In a controller method:

```
@Post
void savePlace(@Valid Place place) {

}
```
